### ISSUE TITLE
Wrap union and intersection types inside arrows in parentheses

### ISSUE THREAD

**Issue by nicolo-ribaudo** [prettier/prettier#4480]
**Prettier 1.12.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAKtgXmwAoBKQgPmwEFsAfbAIQG4QAaECDGAS2gGdkoAIYAnURADuABTEJBKYQDcIPACbsQAI1HCwAazgwAyhj08oAc2QxRAVzgcL-OKJjTdlgLbDkAM2EAGxcOACt+AA9GXQMjY2EvOAAZCzh-IJCQM1EXUWRtYS00QOhNDFELGAB1dRgAC2QADgAGDnKIFyrdDHzyuFylNI5ROABHOx4Rj2FvXyQA4McQFy8eG3sl-gtLQLgARTsIeHTFjhhCmrV65AAmM90eQO2AYQgvH3yoaCGQOxdcQoKBYuAC+IKAA)

**Input:**
```jsx
type T = () => A | B;
```

**Output:**
```jsx
type T = () => A | B;

```

**Expected behavior:**

Maybe prettier whould add parentheses around `A | B`, to make it easier to understand what is inside of the function and what is not. I'm personally not in favour of this change, but I'm opening this issue to continue the discussion started on https://github.com/prettier/prettier/pull/4475#discussion_r187808437.

**Comment by hawkrives** [comment]
I'll go ahead and copy over my comment:

> My personal anecdote: I can't remember the operator precedence for `|`, so I actually thought that the type was `(?() => A) | (B)`. So parenthesis would be helpful for me.

I feel like this can draw parallels with the boolean operators, where `a || b && c` turns into `a || (b && c)` – that is, boolean operators were determined to be a point of confusion, so parentheses are printed around them. I feel like Flow's `|` and `&` may have the same issue of confusion, so my vote is for adding parentheses.

**Comment by nicolo-ribaudo** [comment]
Then should we also add parentheses to `() => a || b`?

**Comment by j-f1** [comment]
Maybe

**Comment by duailibe** [comment]
I don't think we need this.. it's pretty clear to me in an assignment that everything after `=>` is the return "part"

**Comment by hawkrives** [comment]
@duailibe Thats fair. I'll try to remember that. 

Given @nicolo-ribaudo's question and thought, I'd be fine with not changing the output of Prettier for either arrow functions returning boolean value _or_ arrow function flow types. 

I'll figure out how to visually parse the this eventually ;-)
